modules/l4remoteiplist/iplist.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2024 SICK AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l4remoteiplist
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+type IpList struct {
+	ipFile            string         // File containing all IPs to be matched, gets continously monitored
+	ipAddresses       []string       // List of currently loaded IP addresses that would matched
+	ctx               *caddy.Context // Caddy context, used to detect when to shut down
+	logger            *zap.Logger
+	reloadNeededMutex sync.Mutex // Mutex to ensure proper concurrent handling of reloads
+	reloadNeeded      bool       // Flag indicating whether a reload of the IPs is needed
+}
+
+// Creates a new IpList, creating the ipFile if it is not present
+func NewIpList(ipFile string, ctx *caddy.Context, logger *zap.Logger) (*IpList, error) {
+	ipList := &IpList{
+		ipFile:       ipFile,
+		ctx:          ctx,
+		logger:       logger,
+		reloadNeeded: true,
+	}
+
+	if !ipList.ipFileExists() {
+		logger.Debug("could not find the file containing the IPs, trying to create it...")
+
+		// Create a new file since it does not exist
+		_, err := os.Create(ipFile)
+		if err != nil {
+			logger.Error("Error creating the IP list", zap.Error(err))
+			return nil, fmt.Errorf("cannot create a new IP list: %v", err)
+		}
+		logger.Debug("list of IP addresses successfully created")
+	}
+
+	return ipList, nil
+}
+
+// Check whether a IP address is currently contained in the IP list
+func (b *IpList) IsMatched(ip string) bool {
+	// First reload the IP list if needed to ensure IPs are always up to date
+	b.reloadNeededMutex.Lock()
+	reloadNeeded := b.reloadNeeded
+	b.reloadNeeded = false
+	b.reloadNeededMutex.Unlock()
+
+	if reloadNeeded {
+		err := b.loadIpAddresses()
+		if err != nil {
+			b.logger.Error("could not load IP addresses")
+		} else {
+			b.logger.Debug("reloaded IP addresses")
+		}
+	}
+
+	for _, listIp := range b.ipAddresses {
+		if listIp == ip {
```

Comment:
You match strings, and it would probably work fine for IPv4. Consider IPv6: fd00::1 is a synonym to fd00:0:0:0:0:0:0:1, but their string representations aren't equal. So I suggest we match IP addresses, not strings.